### PR TITLE
[SPARK-33914][SQL][DOCS] Describe the structure of unified DS v1 and v2 tests

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
Add comments for the unified datasource tests, describe what kind of tests they contain, and put refs to other test suits.

### Why are the changes needed?
To improve code maintenance.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
By running `./dev/scalastyle`.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableAddPartitionSuiteBase.scala [hunk 1]
```diff
@@ -22,6 +22,20 @@ import org.apache.spark.sql.catalyst.analysis.PartitionsAlreadyExistException
...
+ * located in more specific test suites:
+ *
+ *   - V2 table catalog tests:
+ *     [[org.apache.spark.sql.execution.command.v2.AlterTableAddPartitionSuite]]
+ *   - V1 table catalog tests:
+ *     [[org.apache.spark.sql.execution.command.v1.AlterTableAddPartitionSuiteBase]]
+ *     - V1 In-Memory catalog:
+ *       [[org.apache.spark.sql.execution.command.v1.AlterTableAddPartitionSuite]]
+ *     - V1 Hive External catalog:
+ *       [[org.apache.spark.sql.hive.execution.command.AlterTableAddPartitionSuite]]
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableAddPartitionSuiteBase.scala:37]: @dongjoon-hyun It seems that the ref to outer package causes some troubles in generating java docs. I will replace `[[]]` by `''`

**Comment by MaxGekk** [comment]
Not clear how my changes are related to the error:
```
[error] /home/runner/work/spark/spark/mllib/target/java/org/apache/spark/mllib/util/MLlibTestSparkContext.java:10:1:  error: illegal combination of modifiers: public and protected
[error]   protected  class testImplicits {
[error]              ^
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableDropPartitionSuiteBase.scala [hunk 2]
```diff
@@ -21,6 +21,20 @@ import org.apache.spark.sql.{AnalysisException, QueryTest}
...
 
+/**
+ * This base suite contains unified tests for the `ALTER TABLE .. DROP PARTITION` command that
+ * check V1 and V2 table catalogs. The tests that cannot run for all supported catalogs are
+ * located in more specific test suites:
+ *
+ *   - V2 table catalog tests:
+ *     [[org.apache.spark.sql.execution.command.v2.AlterTableDropPartitionSuite]]
+ *   - V1 table catalog tests:
+ *     [[org.apache.spark.sql.execution.command.v1.AlterTableDropPartitionSuiteBase]]
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableDropPartitionSuiteBase.scala:32]: `org.apache.spark.execution.command.AlterTableDropPartitionSuiteBase` ?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableAddPartitionSuiteBase.scala [hunk 3]
```diff
@@ -22,6 +22,20 @@ import org.apache.spark.sql.catalyst.analysis.PartitionsAlreadyExistException
...
 
+/**
+ * This base suite contains unified tests for the `ALTER TABLE .. ADD PARTITION` command that
+ * check V1 and V2 table catalogs. The tests that cannot run for all supported catalogs are
+ * located in more specific test suites:
+ *
+ *   - V2 table catalog tests:
+ *     [[org.apache.spark.sql.execution.command.v2.AlterTableAddPartitionSuite]]
+ *   - V1 table catalog tests:
+ *     [[org.apache.spark.sql.execution.command.v1.AlterTableAddPartitionSuiteBase]]
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableAddPartitionSuiteBase.scala:33]: `org.apache.spark.sql.execution.command.AlterTableAddPartitionSuiteBase`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableDropPartitionSuiteBase.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableDropPartitionSuiteBase.scala:32]: No, I think the ref is correct, see https://github.com/apache/spark/blob/29cca68e9e55fae8389378de6f30d0dfa7a74010/sql/core/src/test/scala/org/apache/spark/sql/execution/command/v1/AlterTableDropPartitionSuite.scala#L23

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableAddPartitionSuiteBase.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableAddPartitionSuiteBase.scala:33]: You pointed out the current trait which contains tests for v1 and v2. `org.apache.spark.sql.execution.command.v1.AlterTableAddPartitionSuiteBase` is the base trait for v1 In-Memory and Hive catalogs.

Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableAddPartitionSuiteBase.scala:33]: Unidoc doesn't work for the referencing in Javadoc with Scala's trait. it should be `` `org.apache.spark.sql.execution.command.v1.AlterTableAddPartitionSuite` ``

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableAddPartitionSuiteBase.scala:33]: Thank you very much @HyukjinKwon . Without your comment, I would spend much time in troubleshooting this.

**Comment by HyukjinKwon** [comment]
+1 I like this comments!

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableDropPartitionSuiteBase.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/command/AlterTableDropPartitionSuiteBase.scala]: My bad~

**Comment by dongjoon-hyun** [comment]
+1, late LGTM. Thank you.

**Pull request closed without merging**